doc/pantry.md

```diff
@@ -0,0 +1,482 @@
+format. Therefore, there is no longer such a thing as "custom
+snapshots," there are simply "snapshots." Pantry snapshots follow the
+same format as Stack 1.6 "custom snapshots."
+
+Snapshots provide a list of packages to use, along with flags,
+ghc-options, and a few other settings. Snapshots may extend any other
+snapshot that can be specified in a `resolver` field. The packages
+specified follow the same syntax mentioned above for
+dependencies. Unlike `extra-deps`, however, no support for local
+directories is available in snapshots to ensure reproducibility.
+
+```yaml
+resolver: lts-8.21 # Inherits GHC version and package set
+compiler: ghc-8.0.1 # Overwrites GHC version in the resolver, optional
+
+name: my-snapshot # User-friendly name
+
+# Additional packages, follows extra-deps syntax
+packages:
+- unordered-containers-0.2.7.1
+- hashable-1.2.4.0
+- text-1.2.2.1
+
+# Override flags, can also override flags in the parent snapshot
+flags:
+  unordered-containers:
+    debug: true
+
+# Packages from the parent snapshot to ignore
+drop-packages:
+- wai-extra
+
+# Packages which should be hidden (affects script command's import
+# parser
+hidden:
+  wai: true
+  warp: false
+
+# Set GHC options for specific packages
+ghc-options:
+  warp:
+  - -O2
+```
+
+If you put this in a `snapshot.yaml` file in the same directory as your project,
+you can now use the custom snapshot like this:
+
+```yaml
+resolver: snapshot.yaml
+```
+
+This is an example of a custom snapshot stored in the filesystem. They are
+assumed to be mutable, so you are free to modify it. We detect that the snapshot
+has changed by hashing the contents of the involved files, and using it to
+identify the snapshot internally. It is often reasonably efficient to modify a
+custom snapshot, due to stack sharing snapshot packages whenever possible.
+
+Running the `stack freeze --snapshot` command yields the following
+output:
+
+```yaml
+flags:
+  unordered-containers:
+    debug: true
+ghc-options:
+  warp:
+  - -O2
+packages:
+- hackage: unordered-containers-0.2.7.1@sha256:7a1ceb6d88c0f16ec417f28dac16f6dc7b10e88fbb536a74d84941ad2f57b74b,4367
+  pantry-tree:
+    size: 1286
+    sha256: 8a8f745cacae3c11a9c6e6c2fcefc95a13d0c153a8e14b4d28485db1b59d9ef3
+- hackage: hashable-1.2.4.0@sha256:33a49b3ea87cc4a0c89a4fd48f19e4807d8c620aff710a048a28cf7d9c9b4620,4271
+  pantry-tree:
+    size: 1325
+    sha256: cb05c31a8ec43f727004e5a6c8e35ff92e0515855a85cb01fa73623683ee4b33
+- hackage: text-1.2.2.1@sha256:1c6ffad395d1674915cc9fda1d3b8f202ddcbfda7c341eb8bd99de67d3283bf9,5724
+  pantry-tree:
+    size: 7376
+    sha256: ac2601c49cf7bc0f5d66b2793eddc8352f51a6ee989980827a0d0d8169700a03
+name: my-snapshot
+hidden:
+  warp: false
+  wai: true
+drop-packages:
+- wai-extra
+compiler: ghc-8.0.1
+resolver:
+  size: 515969
+  url: https://raw.githubusercontent.com/commercialhaskell/stackage-snapshots/master/lts/8/21.yaml
+  sha256: 2ec73d520d3e55cb753eaca11a72a9ce95bd9ba7ccaf16de1150d0130a50a5a1
+```
```

Comment:
Upgrading this looks like a bit of a nightmare — is there a way to undo a freeze?

Reply:
That's an interesting idea. We _could_ provide an `unfreeze` command which removes all of the "unnecessary" information. Personally, I'd rather wait to try this out a bit and see if such a command is necessary. I would imagine this may actually be easier in practice than it first seems.

Reply:
Sure. It's easy enough to work around, and might be a nice issue for a newcomer.